src/number/n_unknown.jl

```diff
@@ -486,29 +507,69 @@ end
 
 end # for (rw, rew) in
 
+CoefficientRingID = Dict{Nemo.Ring, Any}()
 
-function CoefficientRing(R::Nemo.Ring)
-   T = elem_type(R)
+if VERSION >= v"1.7"
+  _can_i_haz_mutable(::Type{T}) where {T} = ismutabletype(T) && ismutabletype(elem_type(T))
+else
+  _can_i_haz_mutable(::Type{T}) where {T} = T.mutable && elem_type(T).mutable
+end
 
-   if VERSION >= v"1.8"
-      ok = ismutabletype(T)
-   else
-      ok = ismutable(R())
-   end
+# insert a dummy type to force some things to happen at compile time
+struct _Dummy{T}
+end
 
-   if R isa Nemo.Field
-      if ok
-         return N_Field{T}(R)
-      else
-         RR = FieldWrapper{typeof(R), elem_type(R)}(R)
-         return N_Field{elem_type(RR)}(RR)
-      end
-   else
-      if ok
-         return N_Ring{T}(R)
-      else
-         RR = RingWrapper{typeof(R), elem_type(R)}(R)
-         return N_Ring{elem_type(RR)}(RR)
-      end
-   end
+_Dummy{T}(x::T) where {T} = x
+
+@generated function _can_i_haz_type(::Type{T}) where {T <: Nemo.Field}
+  if _can_i_haz_mutable(T)
+    return :(_Dummy{T})
+  else
+    S = FieldWrapper{T, elem_type(T)}
+    return :($S)
+  end
+end
+
+@generated function _can_i_haz_type(::Type{T}) where T
```

Comment:
Should this be restricted to Ring? Or perhaps you had a reason to leave it open?
```suggestion
@generated function _can_i_haz_type(::Type{T}) where {T <: Nemo.Ring}
```

Reply:
Oh, in fact, can't you merge this method with the previous `@generated` method? Just adding another `elseif` case shouldn't hurt.

If one was boring and not into genius names, one could then also call this `mutable_field_or_ring_type_wrapper` or some such silly thing

Reply:
I will try to clean it up and use some boring names.